cmd/kubeadm/app/cmd/phases/addons.go

```diff
@@ -0,0 +1,124 @@
+		Aliases: []string{"addons"},
+		Short:   "Install an addon to a Kubernetes cluster.",
+		RunE:    subCmdRunE("addon"),
+	}
+
+	cmd.AddCommand(getAddonsSubCommands()...)
+	return cmd
+}
+
+func EnsureAllAddons(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error {
+	return nil
+}
+
+// getAddonsSubCommands returns sub commands for addons phase
+func getAddonsSubCommands() []*cobra.Command {
+	cfg := &kubeadmapiext.MasterConfiguration{}
+	// Default values for the cobra help text
+	api.Scheme.Default(cfg)
+
+	var cfgPath, kubeConfigFile string
+	var subCmds []*cobra.Command
+
+	subCmdProperties := []struct {
+		use     string
+		short   string
+		cmdFunc func(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error
+	}{
+		{
+			use:     "all",
+			short:   "Install all addons to a Kubernetes cluster",
+			cmdFunc: EnsureAllAddons,
+		},
+		{
+			use:     "kube-dns",
+			short:   "Install kube-dns addon to a Kubernetes cluster.",
+			cmdFunc: dnsaddon.EnsureDNSAddon,
+		},
+		{
+			use:     "kube-proxy",
+			short:   "Install kube-proxy addon to a Kubernetes cluster.",
+			cmdFunc: proxyaddon.EnsureProxyAddon,
+		},
+	}
+
+	for _, properties := range subCmdProperties {
+		// Creates the UX Command
+		cmd := &cobra.Command{
+			Use:   properties.use,
+			Short: properties.short,
+			Run:   runAddonsCmdFunc(properties.cmdFunc, cfg),
+		}
+
+		// Add flags to the command
+		cmd.Flags().StringVar(&kubeConfigFile, "kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use for talking to the cluster")
+		cmd.Flags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+
+		subCmds = append(subCmds, cmd)
+	}
+
+	return subCmds
+}
+
+// runAddonsCmdFunc creates a cobra.Command Run function, by composing the call to the given cmdFunc with necessary additional steps (e.g preparation of input parameters)
+func runAddonsCmdFunc(cmdFunc func(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error, cfg *kubeadmapiext.MasterConfiguration) func(cmd *cobra.Command, args []string) {
+
+	// the following statement build a clousure that wraps a call to a cmdFunc, binding
+	// the function itself with the specific parameters of each sub command.
+	// Please note that specific parameter should be passed as value, while other parameters - passed as reference -
+	// are shared between sub commands and gets access to current value e.g. flags value.
+
+	return func(cmd *cobra.Command, args []string) {
+		if err := validation.ValidateMixedArguments(cmd.Flags()); err != nil {
+			kubeadmutil.CheckErr(err)
+		}
+
+		var kubeConfigFile string
+		internalcfg := &kubeadmapi.MasterConfiguration{}
+		api.Scheme.Convert(cfg, internalcfg, nil)
+		client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
+		kubeadmutil.CheckErr(err)
+		// internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(*cfgPath, cfg)
```

Comment:
why commented out?
Also pass `cfgPath` to this func

Reply:
Whoops. This was a mistake.